### ISSUE TITLE
Allow displaying groups of fields in one View configuration

### ISSUE THREAD

**Issue by zackkatz** [gravityview/GravityView#1173]
## Why this is needed:

Users are creating a View and embedding it multiple times on the same page, to show entries filtered by a different value. For example: a View that shows issues that can be grouped by status.

Allow grouping entries by Status, then looping through each status:

```
[ Top widgets, like search entries widget ]

## Active
[ table output of only entries with "active" value ]

## Pending
[ table output of only entries with "pending" value ]

## Closed
[ table output of only entries with "closed" value ]

[ Bottom widgets, like pagination links widget]
```

This would repeat the body of a View multiple times, one for each Group. The Top and Bottom widgets would only be rendered once.

## Things to consider

- Not sure if this is possible yet with GF_Query. @soulseekah ?
- The value of the group will need to be limited to field types with choices (Radio/Dropdown/Multi-Select)
- The value of each choice should be added to the group as a header (h3?)
- A setting would need to be added to the View Settings > Sort & Filter metabox to enable, then to choose which field

**Comment by soulseekah** [comment]
> Not sure if this is possible yet with GF_Query.

This isn't GROUP BY functionality. This is ORDER BY, which is supported even without GF_Query.

This is mere ordering with BREAKS when the values changes. SQL-wise there's no difference, Entry_Collection-wise there's no difference.

This is a template issue first and foremost really.

We can reuse the renderer and only feed it Entry_Collections broken down by a field value, which we do in PHP, not in SQL code. So we get all the entries, break them in into separate collections and render them, for example.

Or we can create a new template that knows how to break and output h3 and whatnot, start a table etc. Can be plugin, our core doesn't need to be changed for this at all. Unless we're thinking something very fundamental that would output templates in a loop (might get messy).

**Comment by soulseekah** [comment]
## Configuration

```
"field_groups": [ "date_created", "1" ]
```

I suggest this be a setting under Filter & Sort. Requires `GF_Query` (because multisorts).

## Template

The template is called "table-group". It inherits from "table". It is activated automatically if "field_groups" is not empty.

**Comment by soulseekah** [comment]
## Sorting

"Grouping by a field" actually implies sorting by that field. Grouping prepends sorts to any other implicit or explicit sorts for the view.

Edge cases can be had, but you'd have to be doing something very very weird with the configuration. Like grouping and sorting and using some obscure filters. Time will tell.

**Comment by soulseekah** [comment]
## Paging

Paging subgroups is going to get very weird. By default we'll page the whole resultset across all "groups". If the resultset spans multiple pages group headers can end up being on pages other than the first.

**Group A**
entry 1
entry 2
entry 3
**Group B**
entry 4
entry 5
entry 6

If the paging setting is set to 2 only the first 2 entries from Group A are going to be visible. Going to page 2 will show entry 3, Group B header, entry 4. Page 3 will show entry 5 and 6. We can finetune the behavior later, but paging within groups will not be possible (a least in this iteration).

**Comment by soulseekah** [comment]
## Magic

You can further group subgroups together by using a special filter.
It will allow things like grouping by week, by day, by month, by year, even by starting letter.

The `gravityview/template/table/in_group` allows user to remain in the current group, or click over to the next group for any given entry during the output loop.

## Other known limitations

Cannot group by dynamic values. One specific instance of this is the content field.

Cannot output groups in arbitrary order. Since it's a resultset sort we output things as they were queried.

Widgets will only be output once. Obviously.

You will never get the same flexibility and functionality as you would with separate views as users have been doing.

**Comment by soulseekah** [comment]
This is ready for testing on the `feature/groupview` branch.

How to test at the moment:

1. Update the View settings manually:
```
$s = get_post_meta( $view_id, '_gravityview_template_settings', true );
$s['group_fields'] = array( '1', '2' );
update_post_meta( $view_id, '_gravityview_template_settings', $s );
```

2. Make sure to sort (OR multisort, merge the multisort branch) by the group fields (this is important to get correct results, see above).

**Comment by zackkatz** [comment]
@soulseekah This should be a stand-alone Premium View like DIY. Please break it out into its own private repo (let me know if you need me to do this).

**Comment by soulseekah** [comment]
Create the repo, and ping me on it. I'll move the code.

**Comment by soulseekah** [comment]
Moved to https://github.com/gravityview/Grouped-Tables

Needs UI

**Comment by mikerouse** [comment]
Hey guys, sorry to drop in here, but desperate times are upon us. I've set up a charity in my local area to deliver relief and aid during the Coronavirus crisis, and we're doing some kick-ass stuff with GravityView - this plugin has helped us serve over 131 people and counting so far. 

However... grouping would be awesome - why?

One View we have is to help our volunteers run a prescription collection service. We'd love to group the requests by the pharmacy the prescription is being held at so that one person can go and get all the prescriptions from that one place. I have 13+ chemists and at the moment I would need to create a View for each chemist.

Next, I have my town split into districts and volunteers want to browse available jobs by district. I've got 37 districts in the town and would need to add a view for each district.

Next they want to browse by type of request - food requests, shopping requests, walking the dog requests. You get the picture.

For usability everything has been gathered on one single "I need help" form so it's super easy for people aged over 70 or who might otherwise struggle with technology. The collection end of it works well, but maybe we will need to make separate forms over time. 

However, if grouped tables were available that would be so sweet for our needs.

If you want to see us in action we're at https://supportredditch.net/ 
![screenshot-supportredditch net-2020 04 03-23_20_20](https://user-images.githubusercontent.com/2799126/78409533-d3485800-7601-11ea-9360-41548bf8bc09.png)
